### ISSUE TITLE
The Entropy drops to zero during training processing.

### ISSUE THREAD

**Issue by Yeah21** [Gavinwxy/DGCL#10]
Thank you so much for open-source such solid work. At present, I have done a weakly supervised segmentation task on my dataset only containing the target and background. I learned Teacher-Student Network in your paper. However, in the training processing, the unsupervised loss drops to 0. The reason is the entropy is very small. Do you know if this problem is normal? 

```
def power_jaccard_loss(input: torch.Tensor, target: torch.Tensor):
    input_sigmoid = torch.sigmoid(input)
    eps = 1e-6

    iflat = input_sigmoid.flatten()
    tflat = target.flatten()
    intersection = (iflat * tflat).sum()
    denom = (iflat**2 + tflat**2).sum() - (iflat * tflat).sum() + eps

    return 1 - (intersection / denom)

def compute_unsupervised_loss(predict, target, ignore_mask, criterion):
    batch_size, num_class, h, w = predict.shape
    with torch.no_grad():
        weight = batch_size * h * w / torch.sum(ignore_mask != 255)
    loss = weight * criterion(predict[ignore_mask != 255], target[ignore_mask != 255])
    return loss, 1. / weight.item()

def run_training(cfg):
    run_config = {
        'CONFIG_NAME': cfg.NAME,
        'device': device,
        'epochs': cfg.TRAINER.EPOCHS,
        'learning rate': cfg.TRAINER.LR,
        'batch size': cfg.TRAINER.BATCH_SIZE,
    }
    table = {'run config name': run_config.keys(),
             ' ': run_config.values(),
             }
    print(tabulate(table, headers='keys', tablefmt="fancy_grid", ))

    net = networks.create_network(cfg)
    net.to(device)
    optimizer = optim.AdamW(net.parameters(), lr=cfg.TRAINER.LR, weight_decay=0.01)

    net_teacher = networks.create_network(cfg)
    net_teacher.to(device)
    for p in net_teacher.parameters():
        p.requires_grad = False

    criterion = power_jaccard_loss

    # reset the generators
    loader_sup, loader_unsup, loader_val = datasets.get_loader()


    # unpacking cfg
    epochs = cfg.TRAINER.EPOCHS
    save_checkpoints = cfg.SAVE_CHECKPOINTS
    # tracking variables
    global_step = 0
    for epoch in range(1, epochs + 1):
        print(f'\nStarting epoch {epoch}/{epochs}.')

        loader_sup_iter = iter(loader_sup)
        loader_unsup_iter = iter(loader_unsup)

        supervised_loss_set, unsupervised_loss_set, percent_reliable_set, loss_set = [], [], [], []
        progress = (epoch - 2) / (epochs - 2)
        percent_unreliable = 90 * (1 - progress)
        for step in range(len(loader_sup)):
            global_step += 1
            start = timeit.default_timer()
            batch_l = next(loader_sup_iter)
            image_l, label_l = batch_l['x'].to(device), batch_l['y'].to(device)
            batch_u = next(loader_unsup_iter)
            image_u, ignore_mask = batch_u['x'].to(device), batch_u['y'].to(device)
            if epoch < 2:
                net.train()
                optimizer.zero_grad()

                logits = net(image_l)
                supervised_loss, unsupervised_loss = torch.tensor(0).to(device), torch.tensor(0).to(device)
                supervised_loss = criterion(logits, label_l)
                supervised_loss_set.append(supervised_loss.item())
                unsupervised_loss = unsupervised_loss * cfg.CONSISTENCY_TRAINER.LOSS_FACTOR
                unsupervised_loss_set.append(unsupervised_loss.item())
                loss = unsupervised_loss + supervised_loss
                percent_reliable_set.append(0)
                loss_set.append(loss.item())
                loss.backward()
                optimizer.step()
                ema_decay=0
            else:
                if epoch == 2:
                    with torch.no_grad():
                        for t_params, s_params in zip(net_teacher.parameters(), net.parameters()):
                            t_params.data = s_params.data
                num_labeled = len(image_l)
                net_teacher.train()
                with torch.no_grad():
                    pred_u_t = net_teacher(image_u).detach()
                prob = torch.sigmoid(pred_u_t)
                label_u_t = (prob > 0.5).float()
                entropy = -(prob * torch.log(prob + 1e-10) + (1 - prob) * torch.log((1 - prob) + 1e-10))
                thresh = torch.quantile(entropy.flatten(), (100 - percent_unreliable) / 100)
                thresh_mask = entropy.ge(thresh).bool()
                ignore_mask[thresh_mask] = 255

                if random.uniform(0,1) < 0.5:
                    image_u_aug, label_u_aug, ignore_mask_aug = augmentations.generate_unsup_data(
                        image_u, label_u_t.clone(), ignore_mask, mode='cutmix')
                else:
                    image_u_aug, label_u_aug, ignore_mask_aug = image_u, label_u_t.clone(), ignore_mask.clone()
                # image_u_aug, label_u_aug, ignore_mask_aug = image_u, label_u_t.clone(), ignore_mask.clone()
                image_tri = torch.cat((image_l, image_u_aug), dim=0)
                net.train()
                pred_tri_s = net(image_tri)
                pred_l_s = pred_tri_s[:num_labeled]
                supervised_loss = criterion(pred_l_s, label_l.clone())
                pred_u_aug = pred_tri_s[num_labeled:]
                unsupervised_loss, percent_reliable_batch = compute_unsupervised_loss(pred_u_aug.clone(),
                                                                                      label_u_aug.clone(),
                                                                                      ignore_mask_aug, criterion)
                supervised_loss_set.append(supervised_loss.item())
                percent_reliable_set.append(percent_reliable_batch)
                unsupervised_loss = unsupervised_loss * cfg.CONSISTENCY_TRAINER.LOSS_FACTOR
                unsupervised_loss_set.append(unsupervised_loss.item())
                loss = unsupervised_loss + supervised_loss
                loss_set.append(loss.item())
                loss.backward()
                optimizer.step()

                with torch.no_grad():
                    # ema_decay = min(1 - 1 / (global_step - len(loader_sup)+1), 0.999)
                    ema_decay = 0.99
                    for t_params, s_params in zip(net_teacher.parameters(), net.parameters()):
                        t_params.data = (ema_decay * t_params.data + (1 - ema_decay) * s_params.data)
                        # t_params.data = s_params.data
            # logging
            time = timeit.default_timer() - start
            log = f'step:{global_step}  epoch:{epoch}  loss_set:{np.mean(loss_set):.4f}  ' \
                  f'supervised_loss:{np.mean(supervised_loss_set):.4f}  ' \
                  f'unsupervised_loss:{np.mean(unsupervised_loss_set):.4f}  ' \
                  f'percent_reliable:{np.mean(percent_reliable_set):.4f}  time:{time:.4f}  ema_decay:{ema_decay:.4f}'
            if global_step % 10 == 0:
                sys.stdout.write('\r' + log)
            if step == len(loader_sup):
                sys.stdout.write('\n' + log)
        if epoch == 1:
            evaluation.model_evaluation(net, loader_val, device, epoch, global_step)
        else:
            evaluation.model_evaluation(net, loader_val, device, epoch, global_step)
            evaluation.model_evaluation(net_teacher, loader_val, device, epoch, global_step)

        if epoch in save_checkpoints:
            print(f'\nsaving network', flush=True)
            networks.save_checkpoint(net_teacher, optimizer, epoch, global_step, cfg)

            # logs to load network
            evaluation.model_evaluation(net_teacher, loader_val, device, epoch, global_step)```

**Comment by gavinwxy** [comment]
Hi, thanks for your interest in this work. Based on the codes you provide, my guess is that the filtering threshold is too strict for your case, which means only the samples with lowest entropy (in other word, highest confidence) can involve in the training. That's the reason why the loss is too small being supervised by pseudo labels. 

For better understanding,  
<img width="290" alt="image" src="https://github.com/Gavinwxy/DGCL/assets/27850859/e94fca38-c211-451b-9149-ac308195c819">
<img width="580" alt="image" src="https://github.com/Gavinwxy/DGCL/assets/27850859/b2cb6f4c-1226-4787-a2a8-79bc605a96b2">

here, at the beginning of training, you only keep samples with lowest 10% entropy. In your case, maybe only 0 entropy samples will meet the requirement. My suggestion is that you can loose the threshold, let's say, to 50% of samples. Then you might have larger losses in training.

**Comment by Yeah21** [comment]
Thank you very much for your reply and suggestions. I am currently trying different thresholds. I'm a little puzzled, why does the teacher model use train() instead of eval() when making pseudo-labels?

**Comment by gavinwxy** [comment]
That is because we need the BatchNorm layer in teacher model to adapt to input image but not from parameter EMA, a little trick that can boost performance. BatchNorm layer statistics can be very important to performance based on the research in tasks like test-time adaptation.
